### ISSUE TITLE
Using the async features of the driver

### ISSUE THREAD

**Issue by mistahmikey** [impossibl/pgjdbc-ng#162]
I finally was able to get this working a bit, but for some reason, after receiving a few notifications, your event loop stops receiving them altogether.

In Postgresql, I have set up a trigger on a table to call pg_notify whenever rows are inserted into the table.  That appears to be working fine, and as I said, I do see the driver respond to some of the notifications.

I am registering with the driver for notifications within a Wildfly hosted app.  It proved to be a bit tricky to get my hands on a PGConnection from a CDI-injected, Hibernate-based EntityManager to perform the notification registration, and that might be the problem, because for all I know, Wildfly could be closing the connection I used before the session ends.  Is it possible to enable some logging in your driver that might tell me what is happening with the PGConnection?  Or is there some better way to interact with the driver for async notifications from within a web-app?  I have searched the internet, and have found only one partial example of using this feature. In any case, if you could point me to some detailed documentation about how this is supposed to work, I would appreciate it.

**Comment by jesperpedersen** [comment]
If you are inside a transaction you will get the access to the same physical connection as the EM. Then you can use https://github.com/ironjacamar/ironjacamar/blob/1.2/adapters/src/main/java/org/jboss/jca/adapters/jdbc/WrappedConnection.java#L1571

You can enable "spy" on the datasource, and see the SQL performed. Or use another framework like p6spy. Currently there is no SQL level logging in the driver.

Maybe try and use a tool like Byteman to debug the event loop...

**Comment by kdubb** [comment]
Is this a connection pool? It sounds like it would be.  If that's the case each connection would need to invoke LISTEN.  This might explain why you get a few then none.

**Comment by jesperpedersen** [comment]
Yup, http://www.ironjacamar.org/doc/schema/datasources_1_3.xsd

A hack could be to use a non-JTA datasource (jta="false"), then keep the connection open and do LISTEN (it is returned to the pool upon close()), and feed the data into a connection that is used in a JTA context (jta="true").

**Comment by kdubb** [comment]
That was going to be my next suggestion; a dedicated LISTEN connection sourced from the same datasource.

**Comment by jesperpedersen** [comment]
Different datasource definition pointing to the same database, yes

**Comment by mistahmikey** [comment]
Can you elaborate a bit more on how to enable "spy" on the datasource?  I am a bit of a neophyte in the webapp arena.

Here is how I am getting the connection.  I have a SessionScope bean into which I inject an EntityManager that is produced by an ApplicationScope bean.

When I first request the JSF page I wish to display, the Session bean gets created by Wildfly. In the Session bean, I have a @PostConstruct method that gets the "underlying" connection from the injected EntityManager just as you suggested.  In this method, I register a listener with the driver and execute a "LISTEN" on the notification channel.  When the notification method of the listener is called, the payload contains the primary key of the record that was inserted.  I then pass this primary key off to another thread to actually pull the record so it can be used to update the screen (an Icefaces ace:datatable component).

What I am seeing is a fairly random situation where I get updates for a while, sometimes more, sometimes less, but they always stop coming after a few updates.  If I close the browser and start a new session, it repeats the same behavior (but with a different Session bean.) 

Is there any way for me to know when the PGConnection I am using for my notifications is being closed? Or are there other reasons why the notification listener would stop responding to the events?

**Comment by mistahmikey** [comment]
FYI, I did create a JDBC-level connection to the database to guarantee a single, permanent connection to the database would exist during my session for use with your notification scheme.  The behaviour, unfortunately, is the same.

At that point, I decided to drop back and use the Postgresql 9.4 JDBC driver and use their polled implementation.  This required a slight modification of my code (I basically created a notification manager that had the same interface as yours on top of the polled approach, using a thread to do the polling and notification), so the architecture is virtually identical.  This works reliably.

Using your driver, I also occasionally experienced some corruption of the objects I retrieve using JPA.  Again, using my polled implementation, I have yet to see this.

So, it would seem there is something amiss in your driver with respect to the async mechanism, I suspect due to some kind of concurrency synchronization problem.  If there is anything you would like me to try to help you debug this, let me know.
